lib/vtls/schannel.c

```diff
@@ -1576,6 +1610,159 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
   return CURLE_OK;
 }
 
+static CURLcode pkp_pin_peer_pubkey(struct connectdata *conn, int sockindex,
+                                    const char *pinnedpubkey)
+{
+  SECURITY_STATUS status;
+  struct Curl_easy *data = conn->data;
+  struct ssl_connect_data *connssl = &conn->ssl[sockindex];
+  CERT_CONTEXT *pCertContextServer = NULL;
+  const CERT_CHAIN_CONTEXT *pChainContext = NULL;
+  HCRYPTPROV hCryptProv = 0;
+  HCRYPTKEY hCertPubKey = NULL;
+  size_t bloblen, pubkeylen, realpubkeylen;
+  unsigned char *blob = NULL, *pubkey = NULL, *realpubkey = NULL,
+                *spkiHeader = NULL;
+
+  /* Result is returned to caller */
+  CURLcode result = CURLE_SSL_PINNEDPUBKEYNOTMATCH;
+
+  /* if a path wasn't specified, don't pin */
+  if(!pinnedpubkey)
+    return CURLE_OK;
+
+  do {
+    status = s_pSecFn->QueryContextAttributes(&connssl->ctxt->ctxt_handle,
+                                              SECPKG_ATTR_REMOTE_CERT_CONTEXT,
+                                              &pCertContextServer);
+
+    if((status != SEC_E_OK) || (pCertContextServer == NULL)) {
+      failf(data, "schannel: Failed to read remote certificate context: %s",
+            Curl_sspi_strerror(conn, status));
+      break; /* failed */
+    }
+
+
+    if(!CryptAcquireContext(&hCryptProv, NULL, NULL,
+                            /*PROV_RSA_FULL,*/
+                            PROV_RSA_AES,
+                            CRYPT_VERIFYCONTEXT | CRYPT_SILENT)) {
+      failf(data, "schannel: Failed to acquire crypto context: %s",
+            Curl_sspi_strerror(conn, GetLastError()));
+      break; /* failed */
+    }
+
+    /* Get the public key information for the certificate.
+     This works for RSA keys, but secp256r1 and secp384r1 gives
+     this error:
+     Unknown error (0x8009310B) - ASN1 bad tag value met.
+     */
+    if(!CryptImportPublicKeyInfo(
+        hCryptProv,
+        X509_ASN_ENCODING,
+        &pCertContextServer->pCertInfo->SubjectPublicKeyInfo,
+        &hCertPubKey)) {
+      failf(data, "schannel: Failed to import public key info: %s",
+            Curl_sspi_strerror(conn, GetLastError()));
+      break; /* failed */
+    }
+
+    /* export to blob */
+    if(!CryptExportKey(hCertPubKey, 0, PUBLICKEYBLOB, 0, 0, &bloblen)) {
+      failf(data, "schannel: Failed to get public key blob length: %s",
+            Curl_sspi_strerror(conn, GetLastError()));
+      break; /* failed */
+    }
+
+    blob = malloc(bloblen);
+    if(!blob)
+      break; /* failed */
+
+    if(!CryptExportKey(hCertPubKey, 0, PUBLICKEYBLOB, 0, blob, &bloblen)) {
+      failf(data, "schannel: Failed to export public key to blob: %s",
+            Curl_sspi_strerror(conn, GetLastError()));
+      break; /* failed */
+    }
+
+    /* export to der 
+     not sure what second arg should be,
+     X509_PUBLIC_KEY_INFO -- fails
+     RSA_CSP_PUBLICKEYBLOB -- not quite der, missing spki header bytes
+     CNG_RSA_PUBLIC_KEY_BLOB -- fails
```

Comment:
I will take a look into this. You may also find some answers in my WinCNG-based implementation for libssh2. CNG is the successor of crypt32 and requires the use of the bcrypt library. See:  https://github.com/libssh2/libssh2/blob/master/src/wincng.c